fl4health/strategies/feddg_ga_strategy.py

```diff
@@ -0,0 +1,412 @@
+
+        Returns:
+            (NDArrays) the weighted and aggregated results.
+        """
+
+        aggregated_results: Optional[NDArrays] = None
+        for client_proxy, fit_res in results:
+            cid = client_proxy.cid
+
+            # initializing adjustment weights for this client if they don't exist yet
+            if cid not in self.adjustment_weights:
+                assert self.initial_adjustment_weight is not None
+                self.adjustment_weights[cid] = self.initial_adjustment_weight
+
+            # apply adjustment weights
+            weighted_client_parameters = parameters_to_ndarrays(fit_res.parameters)
+            for i in range(len(weighted_client_parameters)):
+                weighted_client_parameters[i] = weighted_client_parameters[i] * self.adjustment_weights[cid]
+
+            # sum weighted parameters
+            if aggregated_results is None:
+                aggregated_results = weighted_client_parameters
+            else:
+                assert len(weighted_client_parameters) == len(aggregated_results)
+                for i in range(len(weighted_client_parameters)):
+                    aggregated_results[i] = aggregated_results[i] + weighted_client_parameters[i]
+
+        assert aggregated_results is not None
+        return aggregated_results
+
+    def update_weights_by_ga(self, server_round: int, cids: List[str]) -> None:
+        """
+        Update the self.adjustment_weights dictionary by calculating the new weights
+        based on the current server round, fit and evaluation metrics.
+
+        Args:
+            server_round: (int) the current server round.
+            cids: (List[str]) the list of client ids that participated in this round.
+        """
+        generalization_gaps = []
+        # calculating local vs global metric difference (generalization gaps)
+        for cid in cids:
+            assert (
+                cid in self.train_metrics and cid in self.evaluation_metrics
+            ), f"{cid} not in {self.train_metrics.keys()} or {self.evaluation_metrics.keys()}"
+
+            assert self.fairness_metric.metric_name is not None
+
+            global_model_metric_value = self.evaluation_metrics[cid][self.fairness_metric.metric_name]
+            local_model_metric_value = self.train_metrics[cid][self.fairness_metric.metric_name]
+            assert isinstance(global_model_metric_value, float) and isinstance(local_model_metric_value, float)
+
+            generalization_gaps.append(global_model_metric_value - local_model_metric_value)
+
+        # Calculating the normalized the generalization gaps
+        generalization_gaps_ndarray = np.array(generalization_gaps)
+        mean_generalization_gap = np.mean(generalization_gaps_ndarray)
+        var_generalization_gaps = generalization_gaps_ndarray - mean_generalization_gap
+        max_var_generalization_gap = np.max(var_generalization_gaps)
```

Comment:
As a small modification, here that they did have in their code and I think makes sense is to take the absolute value of the `var_generalization_gaps` before grabbing the max.

Reply:
Done.